File: packages/browseros-agent/packages/browser-core/src/core/session.ts
import type { ProtocolApi } from '@browseros/cdp-protocol/protocol-api'
import type { CdpConnection } from './connection'
import { Input } from './input/input'
import { Navigation } from './navigation'
import { FrameRegistry } from './observer/frames'
import { Observer } from './observer/observer'
import { PageManager, type PageManagerHooks } from './pages'
import {
  captureScreenshotWithAnnotations,
  type ScreenshotCaptureOptions,
  type ScreenshotCaptureResult,
} from './screenshot'
import { WindowManager } from './windows'

export interface BrowserSessionHooks extends PageManagerHooks {}

/** Coordinates page registry, observation, input, navigation, and raw CDP access. */
export class BrowserSession {
  readonly pages: PageManager
  readonly windows: WindowManager
  private readonly frames: FrameRegistry
  private readonly observers = new Map<number, Observer>()

  constructor(
    private readonly connection: CdpConnection,
    hooks: BrowserSessionHooks = {},
  ) {
    this.frames = new FrameRegistry(connection)
    this.windows = new WindowManager(connection)
    this.pages = new PageManager(connection, {
      ...hooks,
      onSessionAttached: async (session, pageId, sessionId) => {
        await this.frames.registerPage(session, pageId, sessionId)
        await hooks.onSessionAttached?.(session, pageId, sessionId)
      },
    })
    this.connection.Target.on('detachedFromTarget', (params) => {
      if (params.sessionId) this.pages.detachSession(params.sessionId)
    })
  }

  /** Per-page observation (snapshot + diff), created lazily and cached. */
  observe(pageId: number): Observer {
    let observer = this.observers.get(pageId)
    if (!observer) {
      observer = new Observer(this.pages, this.frames, pageId)
      this.observers.set(pageId, observer)
    }
    return observer
  }

  /** The action layer (click/fill/type/...) for a page, sharing its observation refs. */
  input(pageId: number): Input {
    return new Input(this.observe(pageId), this.pages, pageId)
  }

  /** Navigation (url/back/forward/reload) for a page. */
  nav(pageId: number): Navigation {
    return new Navigation(this.pages, pageId)
  }

  /** Captures a page screenshot, optionally overlaying current snapshot refs. */
  async screenshot(
    pageId: number,
    options: ScreenshotCaptureOptions = {},
  ): Promise<ScreenshotCaptureResult> {
    const { session } = await this.pages.getSession(pageId)
    return captureScreenshotWithAnnotations({
      pageSession: session,
      observer: this.observe(pageId),
      options,
    })
  }

  /** Captures only while `pageId` still resolves to the expected CDP target. */
  async screenshotForTarget(
    pageId: number,
    targetId: string,
    options: ScreenshotCaptureOptions = {},
  ): Promise<ScreenshotCaptureResult | null> {
    const selected = this.pages.getInfo(pageId)
    if (!selected || selected.targetId !== targetId) return null
    let resolved: Awaited<ReturnType<PageManager['getSession']>>
    try {
      resolved = await this.pages.getSession(pageId)
    } catch (error) {
      const current = this.pages.getInfo(pageId)
      if (!current || current.targetId !== targetId) return null
      throw error
    }
    const { session, targetId: currentTargetId } = resolved
    if (currentTargetId !== targetId) return null
    const capture = await captureScreenshotWithAnnotations({
      pageSession: session,
      observer: this.observe(pageId),
      options,
    })
    const current = await this.pages.refresh(pageId)
    return current?.targetId === targetId ? capture : null
  }

  /** Raw CDP escape hatch for `run` code, e.g. cdp("Page.navigate", { url }). */
  async cdp(
    method: string,
    params?: Record<string, unknown>,
    sessionId?: string,
  ): Promise<unknown> {
    return this.connection.rawSend(method, params ?? {}, sessionId)
  }

  /** Raw CDP escape hatch that sends already-validated JSON params verbatim. */
  async cdpJson(
    method: string,
    paramsJson: string,
    sessionId?: string,
  ): Promise<unknown> {
    return this.connection.rawSendJson(method, paramsJson, sessionId)
  }

  /** Page-scoped raw CDP for CLI/run callers that start from a BrowserOS page id. */
  async cdpJsonForPage(
    pageId: number,
    method: string,
    paramsJson: string,
  ): Promise<unknown> {
    const { sessionId } = await this.pages.getSession(pageId)
    return this.connection.rawSendJson(method, paramsJson, sessionId)
  }

  isConnected(): boolean {
    return this.connection.isConnected()
  }

  /** Exposes the typed root CDP domains for server-side lifecycle integrations. */
  get protocol(): ProtocolApi {
    return this.connection
  }

  /** Identifies reconnects so lifecycle consumers can rebuild cached browser state. */
  connectionEpoch(): number {
    return this.connection.connectionEpoch()
  }

  async dispose(): Promise<void> {}
}

File: packages/browseros-agent/crates/browseros-core/src/session.rs
use crate::{
    CoreError, PageId, TargetId,
    connection::CdpConnection,
    frames::FrameRegistry,
    input::Input,
    navigation::Navigation,
    observer::Observer,
    page_signals::PageSignals,
    pages::{PageManager, PageManagerHooks},
    screenshot::{
        ScreenshotCaptureOptions, ScreenshotCaptureResult, capture_screenshot_with_annotations,
    },
    windows::WindowManager,
};
use browseros_cdp::CdpEvent;
use serde_json::Value;
use std::{collections::HashMap, sync::Arc};
use tokio::sync::{Mutex, broadcast, mpsc};

#[derive(Clone, Default)]
pub struct BrowserSessionHooks {
    pub page_manager: PageManagerHooks,
}

pub struct BrowserSession {
    connection: Arc<dyn CdpConnection>,
    pub pages: Arc<PageManager>,
    pub page_signals: Arc<PageSignals>,
    pub windows: Arc<WindowManager>,
    frames: Arc<FrameRegistry>,
    observers: Mutex<HashMap<PageId, Arc<Observer>>>,
}

impl BrowserSession {
    #[must_use]
    pub fn new(connection: Arc<dyn CdpConnection>, hooks: BrowserSessionHooks) -> Arc<Self> {
        let frames = FrameRegistry::new(connection.clone());
        let page_signals = PageSignals::new(connection.clone());
        let frame_hook = {
            let frames = frames.clone();
            let page_signals = page_signals.clone();
            let user_hook = hooks.page_manager.on_session_attached.clone();
            Arc::new(
                move |session: crate::ProtocolSession,
                      page_id: PageId,
                      session_id: crate::SessionId| {
                    let frames = frames.clone();
                    let page_signals = page_signals.clone();
                    let user_hook = user_hook.clone();
                    Box::pin(async move {
                        page_signals.attach_page(page_id.clone(), session_id.clone());
                        frames
                            .register_page(session.clone(), page_id.clone(), session_id.clone())
                            .await?;
                        if let Some(user_hook) = user_hook {
                            user_hook(session, page_id, session_id).await?;
                        }
                        Ok(())
                    })
                        as futures_util::future::BoxFuture<'static, Result<(), CoreError>>
                },
            )
        };
        let detach_hook = {
            let page_signals = page_signals.clone();
            let user_hook = hooks.page_manager.on_page_detached.clone();
            Arc::new(move |page_id: PageId| {
                page_signals.forget_page(&page_id);
                if let Some(user_hook) = &user_hook {
                    user_hook(page_id);
                }
            })
        };
        let page_hooks = PageManagerHooks {
            on_session_attached: Some(frame_hook),
            on_page_detached: Some(detach_hook),
        };
        let pages = Arc::new(PageManager::new(connection.clone(), page_hooks));
        let windows = Arc::new(WindowManager::new(connection.clone()));
        let session = Arc::new(Self {
            connection: connection.clone(),
            pages,
            page_signals,
            windows,
            frames,
            observers: Mutex::new(HashMap::new()),
        });
        Self::spawn_detach_listener(session.clone());
        session
    }

    pub async fn observe(&self, page_id: PageId) -> Arc<Observer> {
        let mut observers = self.observers.lock().await;
        if let Some(observer) = observers.get(&page_id) {
            return observer.clone();
        }
        let observer = Arc::new(Observer::new(
            self.pages.clone(),
            self.frames.clone(),
            page_id.clone(),
        ));
        observers.insert(page_id, observer.clone());
        observer
    }

    pub async fn input(&self, page_id: PageId) -> Input {
        Input::new(
            self.observe(page_id.clone()).await,
            self.pages.clone(),
            page_id,
        )
    }

    #[must_use]
    pub fn nav(&self, page_id: PageId) -> Navigation {
        Navigation::new(self.pages.clone(), self.page_signals.clone(), page_id)
    }

    pub async fn screenshot(
        &self,
        page_id: PageId,
        options: ScreenshotCaptureOptions,
    ) -> Result<ScreenshotCaptureResult, CoreError> {
        let page = self.pages.get_session(page_id.clone()).await?;
        capture_screenshot_with_annotations(page.session, self.observe(page_id).await, options)
            .await
    }

    /// Captures only while the reusable page id still names the expected target incarnation.
    /// The checks bracket CDP capture so a rebind can neither capture the replacement nor
    /// publish a result for the old target after it closes.
    pub async fn screenshot_for_target(
        &self,
        page_id: PageId,
        target_id: &TargetId,
        options: ScreenshotCaptureOptions,
    ) -> Result<Option<ScreenshotCaptureResult>, CoreError> {
        let live = self.pages.refresh(page_id.clone()).await?;
        if live
            .as_ref()
            .is_none_or(|page| page.target_id.as_str() != target_id.as_str())
        {
            return Ok(None);
        }
        let page = self.pages.get_session(page_id.clone()).await?;
        if page.target_id.as_str() != target_id.as_str() {
            return Ok(None);
        }
        let capture = capture_screenshot_with_annotations(
            page.session,
            self.observe(page_id.clone()).await,
            options,
        )
        .await?;
        let live = self.pages.refresh(page_id).await?;
        if live
            .as_ref()
            .is_none_or(|page| page.target_id.as_str() != target_id.as_str())
        {
            return Ok(None);
        }
        Ok(Some(capture))
    }

    pub async fn cdp(
        &self,
        method: &str,
        params: Value,
        session_id: Option<&crate::SessionId>,
    ) -> Result<Value, CoreError> {
        self.connection
            .send(method, params, session_id)
            .await
            .map_err(CoreError::from)
    }

    pub async fn cdp_json(
        &self,
        method: &str,
        params_json: &str,
        session_id: Option<&crate::SessionId>,
    ) -> Result<String, CoreError> {
        self.connection
            .send_raw_json(method, params_json, session_id)
            .await
            .map_err(CoreError::from)
    }

    pub async fn cdp_json_for_page(
        &self,
        page_id: PageId,
        method: &str,
        params_json: &str,
    ) -> Result<String, CoreError> {
        let page = self.pages.get_session(page_id).await?;
        self.connection
            .send_raw_json(method, params_json, Some(&page.session_id))
            .await
            .map_err(CoreError::from)
    }

    #[must_use]
    pub fn cdp_events(&self) -> broadcast::Receiver<CdpEvent> {
        self.connection.events()
    }

    /// See [`crate::connection::CdpConnection::events_targeted`].
    #[must_use]
    pub fn cdp_events_targeted(&self, method: &str) -> mpsc::UnboundedReceiver<CdpEvent> {
        self.connection.events_targeted(method)
    }

    #[must_use]
    pub fn is_connected(&self) -> bool {
        self.connection.is_connected()
    }

    pub async fn dispose(&self) {}

    fn spawn_detach_listener(session: Arc<Self>) {
        let mut events = session.connection.events();
        tokio::spawn(async move {
            loop {
                let Ok(event) = events.recv().await else {
                    break;
                };
                if event.method == "Target.detachedFromTarget" {
                    handle_detached_event(&session, event).await;
                }
            }
        });
    }
}

async fn handle_detached_event(session: &BrowserSession, event: CdpEvent) {
    let session_id = event
        .params
        .get("sessionId")
        .and_then(Value::as_str)
        .map(crate::SessionId::from);
    if let Some(session_id) = session_id {
        session.page_signals.detach_session(&session_id);
        session.pages.detach_session(&session_id).await;
    }
}
